distsql/request_builder.go

```diff
@@ -139,25 +141,54 @@ func (builder *RequestBuilder) SetStreaming(streaming bool) *RequestBuilder {
 }
 
 // TableRangesToKVRanges converts table ranges to "KeyRange".
-func TableRangesToKVRanges(tid int64, ranges []*ranger.NewRange) []kv.KeyRange {
+func TableRangesToKVRanges(tid int64, ranges []*ranger.NewRange, fb *statistics.QueryFeedback) []kv.KeyRange {
+	if fb == nil || fb.Hist() == nil {
+		return tableRangesToKVRangesWithoutSplit(tid, ranges)
+	}
+	ranges = fb.Hist().SplitRange(ranges)
 	krs := make([]kv.KeyRange, 0, len(ranges))
+	newRanges := make([]*ranger.NewRange, 0, len(ranges))
 	for _, ran := range ranges {
-		var low, high []byte
-		low = codec.EncodeInt(nil, ran.LowVal[0].GetInt64())
-		high = codec.EncodeInt(nil, ran.HighVal[0].GetInt64())
-		if ran.LowExclude {
-			low = []byte(kv.Key(low).PrefixNext())
-		}
-		if !ran.HighExclude {
-			high = []byte(kv.Key(high).PrefixNext())
-		}
+		low, high := encodeHandleKey(ran)
 		startKey := tablecodec.EncodeRowKey(tid, low)
 		endKey := tablecodec.EncodeRowKey(tid, high)
 		krs = append(krs, kv.KeyRange{StartKey: startKey, EndKey: endKey})
+
+		// When updating the query feedback, we need to use the encoded form to binary search the corresponding index,
+		// but when storing the feedback we need to get the origin int value, so here we use one datum to store both the
+		// origin and encoded value.
+		r := &ranger.NewRange{LowVal: []types.Datum{ran.LowVal[0]}, HighVal: []types.Datum{ran.HighVal[0]}}
```

Comment:
A datum should have only one type and one value.
It's not supposed to be used this way.